### PR TITLE
Windows file association

### PR DESCRIPTION
**Pull request by Ryex**
This is based off of #229 so that would be needed. This adds the same .mrpack and .zip registration in the windows registry when installing with the NISI installer.
In theory it cleans up after the windows shell when uninstalled.

Needs someone to test the windows build of the installer

### CODE REVIEW

**Comment by TayouVR** [comment]
I have confirmed this is working.

-----

I think we should also have .reg files for scoop, choco and winget users to add and remove these registry keys. I have no idea how its handled in other packages, but I have seen packages like the windows terminal on scoop come with this .reg file:
```
Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\Windows Terminal]
@="Open in Windows &Terminal"
"Icon"="W:\\scoop\\apps\\windows-terminal\\current\\wt.exe"
[HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\Windows Terminal\command]
@="\"W:\\scoop\\apps\\windows-terminal\\current\\wt.exe\" -d \"%V\""

[HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\Windows Terminal]
@="Open in Windows &Terminal"
"Icon"="W:\\scoop\\apps\\windows-terminal\\current\\wt.exe"
[HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\Windows Terminal\command]
@="\"W:\\scoop\\apps\\windows-terminal\\current\\wt.exe\" -d \"%V\""
```
Given it has a fixed path that would indicate it is generated at install time... So maybe this is up to the respective package maintainers?

File: cmake/MacOSXBundleInfo.plist.in
Comment (piperswe) [inline cmake/MacOSXBundleInfo.plist.in]: Indentation here is a little weird

Comment (Ryex) [reply cmake/MacOSXBundleInfo.plist.in]: these commits are part of #229 so theses should probably be addressed there.

**Comment by txtsd** [comment]
> ```
> Windows Registry Editor Version 5.00
> 
> [HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\Windows Terminal]
> @="Open in Windows &Terminal"
> "Icon"="W:\\scoop\\apps\\windows-terminal\\current\\wt.exe"
> [HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\Windows Terminal\command]
> @="\"W:\\scoop\\apps\\windows-terminal\\current\\wt.exe\" -d \"%V\""
> 
> [HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\Windows Terminal]
> @="Open in Windows &Terminal"
> "Icon"="W:\\scoop\\apps\\windows-terminal\\current\\wt.exe"
> [HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\Windows Terminal\command]
> @="\"W:\\scoop\\apps\\windows-terminal\\current\\wt.exe\" -d \"%V\""
> ```

What does this do?

**Comment by Ryex** [comment]
> What does this do?

These specific keys 
```
HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\<name>
HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\Background\shell\<name>\command
HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\<name>
HKEY_CURRENT_USER\SOFTWARE\Classes\Directory\shell\<name>\command
```
create a windows shell entry on directory's right click menu that display the provided description and run the provided command. They would not be relevant to prism.

These are the keys that would need to be created for a portable version of Prism

```
[HKEY_CURRENT_USER\Software\Classes\${APP_ID}]
@="${DESCRIPTION}"
[HKEY_CURRENT_USER\Software\Classes\${APP_ID}\DefaultIcon]
@="${ICON}"
[HKEY_CURRENT_USER\Software\Classes\${APP_ID}\shell]
@="open"                  
[HKEY_CURRENT_USER\Software\Classes\${APP_ID}\shell\open]
@="${COMMANDTEXT}"
[HKEY_CURRENT_USER\Software\Classes\${APP_ID}\shell\open\command]
@="${COMMAND}"

[HKEY_CURRENT_USER\Software\Classes\Applications\${APP_EXE}\shell\open\command]
@="${COMMAND}"
[HKEY_CURRENT_USER\Software\Classes\Applications\${APP_EXE}]
"FriendlyAppName"= "${APP_NAME}"

[HKEY_CURRENT_USER\Software\Classes\Applications\${APP_EXE}\Capabilities]
"ApplicationDescription"="${DESCRIPTION}"
[HKEY_CURRENT_USER\Software\RegisteredApplications]
${APP_NAME}="Software\Classes\Applications\${APP_EXE}\Capabilities"


[HKEY_CURRENT_USER\Software\Classes\.zip\OpenWithList]
"${APP_EXE}"=""
[HKEY_CURRENT_USER\Software\Classes\.zip\OpenWithProgids]
"${APP_ID}"=""

[HKEY_CURRENT_USER\Software\Classes\.mrpack\OpenWithList]
"${APP_EXE}"=""
[HKEY_CURRENT_USER\Software\Classes\.mrpack\OpenWithProgids]
"${APP_ID}"=""

[HKEY_CURRENT_USER\Software\Classes\Applications\${APP_EXE}\Capabilities\FileAssociations]
".mrpack"="${APP_ID}"
[HKEY_CURRENT_USER\Software\Classes\Applications\${APP_EXE}\Capabilities\FileAssociations]
".zip"="${APP_ID}"
```
This would be the install `.reg` file. the reverse file would need to be created to delete the keys.
The variables would, for the most part, need to be replaced at install time.
```
!define APP_ID "@Launcher_CommonName@.App"
!define APP_EXE "@Launcher_APP_BINARY_NAME@.exe"
!define APP_ICON "$INSTDIR\${APP_EXE},0"
!define APP_DESCRIPTION "@Launcher_DisplayName@"
!define APP_NAME "@Launcher_DisplayName@"
!define APP_CMD_TEXT "Prism Launcher instance"
```

there are some additional keys the shell creates when these keys are put into use that would also need to be cleaned up.
```
[HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Search\JumplistData] 
 "$INSTDIR\${APP_EXE}"=-
[HKEY_CURRENT_USER\Software\Classes\Local Settings\Software\Microsoft\Windows\Shell\MuiCache]
"$INSTDIR\${APP_EXE}.FriendlyAppName"=-
[HKEY_CURRENT_USER\Software\Classes\Local Settings\Software\Microsoft\Windows\Shell\MuiCache]
"$INSTDIR\${APP_EXE}.ApplicationCompany"=-
[HKEY_CURRENT_USER\Software\Microsoft\Windows\ShellNoRoam\MUICache]
"$INSTDIR\${APP_EXE}"=-
[HKEY_CURRENT_USER\Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Compatibility Assistant\Store]
"$INSTDIR\${APP_EXE}"=-
```

**Comment by TayouVR** [comment]
I looked into the scoop situation, there are post-install hooks, from which I can configure a .reg file. I suppose I will do that for the next release after this is merged
I have no idea about ~~winget or~~ choco..
As pointed out, winget runs the installer

(we put .reg files in the /scripts dir or the bucket, then use powershell "Replace" to put the correct paths)
(for myself or anyone working on the scoop package, see:
https://github.com/ScoopInstaller/Extras/tree/master/bucket/windows-terminal.json#L6
https://github.com/ScoopInstaller/Extras/tree/master/bucket/windows-terminal.json#L28
https://github.com/ScoopInstaller/Extras/tree/master/scripts/windows-terminal)

**Comment by Scrumplex** [comment]
The other PR was merged!

**Comment by Ryex** [comment]
> The other PR was merged!

Re-based on develop o7

**Comment by DioEgizio** [comment]
> I looked into the scoop situation, there are post-install hooks, from which I can configure a .reg file. I suppose I will do that for the next release after this is merged I have no idea about winget or choco..
> 
> (we put .reg files in the /scripts dir or the bucket, then use powershell "Replace" to put the correct paths) (for myself or anyone working on the scoop package, see: [ScoopInstaller/Extras@`master`/bucket/windows-terminal.json#L6](https://github.com/ScoopInstaller/Extras/tree/master/bucket/windows-terminal.json?rgh-link-date=2022-10-27T15%3A04%3A38Z#L6) [ScoopInstaller/Extras@`master`/bucket/windows-terminal.json#L28](https://github.com/ScoopInstaller/Extras/tree/master/bucket/windows-terminal.json?rgh-link-date=2022-10-27T15%3A04%3A38Z#L28) [ScoopInstaller/Extras@`master`/scripts/windows-terminal](https://github.com/ScoopInstaller/Extras/tree/master/scripts/windows-terminal?rgh-link-date=2022-10-27T15%3A04%3A38Z))

winget uses the installer, so it's not an issue

**Review by Scrumplex — APPROVED** [review]

File: program_info/win_install.nsi.in
Comment (DioEgizio) [inline program_info/win_install.nsi.in:315]: technically, they're modpack formats

File: program_info/win_install.nsi.in
Comment (DioEgizio) [inline program_info/win_install.nsi.in]: technically, they're modpack formats

Comment (Ryex) [reply program_info/win_install.nsi.in]: they are, but since this lets you open them and import them "as" an instance I'm not sure what else to name the command.

Comment (DioEgizio) [reply program_info/win_install.nsi.in]: Minecraft modpack maybe

Comment (Ryex) [reply program_info/win_install.nsi.in]: changed.

**Review by DioEgizio — APPROVED** [review]

**Pull request merged**
